orbit/pkg/execuser/execuser_linux.go

```diff
@@ -247,7 +277,71 @@ func parseIDOutput(s string) (int64, error) {
 
 var whoLineRegexp = regexp.MustCompile(`(\w+)\s+(:\d+)\s+`)
 
-func getUserDisplay(user string, opts eopts) (string, error) {
+type guiSessionType int
+
+const (
+	guiSessionTypeX11 guiSessionType = iota + 1
+	guiSessionTypeWayland
+)
+
+func (s guiSessionType) String() string {
+	if s == guiSessionTypeX11 {
+		return "x11"
+	}
+	return "wayland"
+}
+
+// getUserDisplaySessionType returns the display session type (X11 or Wayland) of the given user.
+func getUserDisplaySessionType(uid string) (guiSessionType, error) {
+	cmd := exec.Command("loginctl", "show-user", uid, "-p", "Display", "--value")
+	var stdout bytes.Buffer
+	cmd.Stdout = &stdout
+	if err := cmd.Run(); err != nil {
+		return 0, fmt.Errorf("run 'loginctl' to get user GUI session: %w", err)
+	}
+	guiSessionID := strings.TrimSpace(stdout.String())
+	if guiSessionID == "" {
+		return 0, errors.New("empty GUI session")
+	}
+	cmd = exec.Command("loginctl", "show-session", guiSessionID, "-p", "Type", "--value")
+	stdout.Reset()
+	cmd.Stdout = &stdout
+	if err := cmd.Run(); err != nil {
+		return 0, fmt.Errorf("run 'loginctl' to get user GUI session type: %w", err)
+	}
+	guiSessionType := strings.TrimSpace(stdout.String())
+	switch guiSessionType {
+	case "":
+		return 0, errors.New("empty GUI session type")
+	case "x11":
+		return guiSessionTypeX11, nil
+	case "wayland":
+		return guiSessionTypeWayland, nil
+	default:
+		return 0, fmt.Errorf("unknown GUI session type: %q", guiSessionType)
+	}
+}
+
+// getUserWaylandDisplay returns the value to set on WAYLAND_DISPLAY for the given user.
+func getUserWaylandDisplay(uid string) (string, error) {
+	matches, err := filepath.Glob("/run/user/" + uid + "/wayland-*")
+	if err != nil {
+		return "", fmt.Errorf("list wayland socket files: %w", err)
+	}
+	sort.Slice(matches, func(i, j int) bool {
+		return matches[i] < matches[j]
+	})
+	for _, match := range matches {
+		if strings.HasSuffix(match, ".lock") {
+			continue
+		}
+		return filepath.Base(match), nil
```

Comment:
This seems slightly brittle, in the edge case where anything else besides sockets and lockfiles ends up in here, so we could take the extra step of looking at the file stats to make sure it's a socket, but again probably an edge case.